dot/core/consensus.go

```diff
@@ -0,0 +1,241 @@
+		"epoch", s.epochNumber,
+	)
+
+	return bs, nil
+}
+
+// handleBlockDigest checks if the provided header is the block header for
+// the first block of the current epoch, finds and decodes the consensus digest
+// item from the block digest, and then sets the epoch data for the next epoch
+func (s *Service) handleBlockDigest(header *types.Header) (err error) {
+
+	// check if block header digest items exist
+	if header.Digest == nil || len(header.Digest) == 0 {
+		return fmt.Errorf("header digest is not set")
+	}
+
+	// declare digest item
+	var item types.DigestItem
+
+	// decode each digest item and check its type
+	for _, digest := range header.Digest {
+		item, err = types.DecodeDigestItem(digest)
+		if err != nil {
+			return err
+		}
+
+		// check if digest item is consensus digest type
+		if item.Type() == types.ConsensusDigestType {
+
+			// cast decoded consensus digest item to conensus digest type
+			consensusDigest, ok := item.(*types.ConsensusDigest)
+			if !ok {
+				return errors.New("failed to cast DigestItem to ConsensusDigest")
+			}
+
+			err = s.handleConsensusDigest(header, consensusDigest)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// handleConsensusDigest checks if the first block has been set for the current
+// epoch, if first block has not been set, assumes first block and sets epoch
+// data for the next epoch using NextEpochDescriptor in the ConsensusDigest
+func (s *Service) handleConsensusDigest(header *types.Header, digest *types.ConsensusDigest) error {
+
+	// check if first block has been set for current epoch
+	if s.firstBlock != nil {
+		return fmt.Errorf("first block already set for current epoch")
+	}
+
+	// check if block is from current epoch
+	err := s.verifyCurrentEpoch(header.Hash())
+	if err != nil {
+		return fmt.Errorf("block is not from current epoch: %s", err)
+	}
+
+	// initialize epoch data interface for next epoch
+	nextEpochData := new(babe.NextEpochDescriptor)
+
+	// decode consensus digest data for next epoch
+	err = nextEpochData.Decode(digest.Data)
+	if err != nil {
+		return err
+	}
+
+	// set epoch data for next epoch
+	err = s.bs.SetEpochData(nextEpochData)
+	if err != nil {
+		return err
+	}
+
+	// set first block in current epoch
+	s.firstBlock = header.Number
+
+	return nil
+}
+
+// getBlockEpoch gets the epoch number using the provided block hash
+func (s *Service) getBlockEpoch(hash common.Hash) (epoch uint64, err error) {
+
+	// get slot number to determine epoch number
+	slot, err := s.blockState.GetSlotForBlock(hash)
+	if err != nil {
+		return epoch, fmt.Errorf("failed to get slot from block hash: %s", err)
+	}
+
+	if slot != 0 {
+		// epoch number = (slot - genesis slot) / epoch length
+		epoch = (slot - 1) / 6 // TODO: use epoch length from babe or core config
+	}
+
+	return epoch, nil
+}
+
+// verifyCurrentEpoch verifies the provided block hash is from current epoch
+func (s *Service) verifyCurrentEpoch(hash common.Hash) error {
```

Comment:
this function is a bit confusing, it doesn't verify the current epoch, it checks if a block is in the current epoch. could you rename it? 

also, the function should return a `bool` to say whether it is in the epoch or not. `error` should be reserved for something actually going wrong

Reply:
can do, will fix